crates/completion/src/copilot_chat.rs

```diff
@@ -0,0 +1,336 @@
+
+        CopilotChatRequest::new(
+            model,
+            request
+                .messages
+                .into_iter()
+                .map(|msg| ChatMessage {
+                    role: match msg.role {
+                        Role::User => CopilotChatRole::User,
+                        Role::Assistant => CopilotChatRole::Assistant,
+                        Role::System => CopilotChatRole::System,
+                    },
+                    content: msg.content,
+                })
+                .collect(),
+        )
+    }
+}
+
+struct AuthenticationPrompt {
+    _subscription: Option<Subscription>,
+}
+
+impl AuthenticationPrompt {
+    pub fn new(cx: &mut AppContext) -> Self {
+        fn update_auth_information(cx: &mut AppContext) {
+            cx.spawn(|mut cx| async move {
+                let oauth_token = copilot_chat::fetch_copilot_oauth_token().await?;
+                cx.update_global::<CompletionProvider, _>(|provider, _cx| {
+                    provider.update_current_as::<_, CopilotChatCompletionProvider>(|provider| {
+                        provider.oauth_token = Some(oauth_token.clone());
+                    });
+                })
+            })
+            .detach_and_log_err(cx);
+        }
+
+        let sub = match Copilot::global(cx) {
+            Some(copilot) => {
+                // Fetch the token if the user is already signed in.
+                match copilot.read(cx).status() {
+                    Status::Authorized => update_auth_information(cx),
+                    _ => {}
+                }
+
+                // Subscribe to copilot status changes so we get the token when the user signs in.
+                Some(
+                    cx.observe(&copilot, |model, cx| match model.read(cx).status() {
+                        Status::Authorized => update_auth_information(cx),
+                        Status::SignedOut => {
+                            cx.update_global::<CompletionProvider, _>(|provider, _cx| {
+                                provider.update_current_as::<_, CopilotChatCompletionProvider>(
+                                    |provider| {
+                                        provider.oauth_token = None;
+                                        provider.api_key = None;
+                                        provider.api_key_expiry = None;
+                                    },
+                                );
+                            })
+                        }
+                        _ => {}
+                    }),
+                )
+            }
+            None => None,
+        };
+
+        Self { _subscription: sub }
+    }
+}
+
+impl Render for AuthenticationPrompt {
+    fn render(&mut self, _cx: &mut ViewContext<Self>) -> impl IntoElement {
+        const LABEL: &str =
+            "To use the assistant panel or inline assistant, you must login to Github Copilot. Your Github account must have an active Copilot Chat subscription.";
```

Comment:
"GitHub" should always be spelled with an uppercase "H" in user-facing text:

```suggestion
            "To use the assistant panel or inline assistant, you must login to GitHub Copilot. Your GitHub account must have an active Copilot Chat subscription.";
```

Reply:
Ah, I did not know that. Fixed it!